usr/lib/common/key.c

```diff
@@ -3185,6 +3185,335 @@ CK_BBOOL kea_priv_check_exportability(CK_ATTRIBUTE_TYPE type)
+    *(CK_KEY_TYPE *) type_attr->pValue = CKK_IBM_PQC_DILITHIUM;
+
+    keyform_attr->type = CKA_IBM_DILITHIUM_KEYFORM;
+    keyform_attr->ulValueLen = sizeof(CK_ULONG);
+    keyform_attr->pValue = (CK_BYTE *) keyform_attr + sizeof(CK_ATTRIBUTE);
+    *(CK_ULONG *) keyform_attr->pValue = 1;
+
+    rho_attr->type = CKA_IBM_DILITHIUM_RHO;
+    rho_attr->ulValueLen = 0;
+    rho_attr->pValue = NULL;
+
+    seed_attr->type = CKA_IBM_DILITHIUM_SEED;
+    seed_attr->ulValueLen = 0;
+    seed_attr->pValue = NULL;
+
+    tr_attr->type = CKA_IBM_DILITHIUM_TR;
+    tr_attr->ulValueLen = 0;
+    tr_attr->pValue = NULL;
+
+    s1_attr->type = CKA_IBM_DILITHIUM_S1;
+    s1_attr->ulValueLen = 0;
+    s1_attr->pValue = NULL;
+
+    s2_attr->type = CKA_IBM_DILITHIUM_S2;
+    s2_attr->ulValueLen = 0;
+    s2_attr->pValue = NULL;
+
+    t0_attr->type = CKA_IBM_DILITHIUM_T0;
+    t0_attr->ulValueLen = 0;
+    t0_attr->pValue = NULL;
+
+    t1_attr->type = CKA_IBM_DILITHIUM_T1;
+    t1_attr->ulValueLen = 0;
+    t1_attr->pValue = NULL;
+
+    template_update_attribute(tmpl, type_attr);
+    template_update_attribute(tmpl, keyform_attr);
+    template_update_attribute(tmpl, rho_attr);
+    template_update_attribute(tmpl, seed_attr);
+    template_update_attribute(tmpl, tr_attr);
+    template_update_attribute(tmpl, s1_attr);
+    template_update_attribute(tmpl, s2_attr);
+    template_update_attribute(tmpl, t0_attr);
+    template_update_attribute(tmpl, t1_attr);
+
+    return CKR_OK;
+}
+
+// dilithium_publ_check_required_attributes()
+//
+CK_RV dilithium_publ_check_required_attributes(TEMPLATE *tmpl, CK_ULONG mode)
+{
+
+    CK_ATTRIBUTE *attr = NULL;
+    CK_BBOOL found;
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_SEED, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_T1, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_KEYFORM, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    return publ_key_check_required_attributes(tmpl, mode);
+}
+
+// dilithium_priv_check_required_attributes()
+//
+CK_RV dilithium_priv_check_required_attributes(TEMPLATE *tmpl, CK_ULONG mode)
+{
+    CK_ATTRIBUTE *attr = NULL;
+    CK_BBOOL found;
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_RHO, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
```

Comment:
so in case mode != MODE_CREATE, its okay if the required attribute is not found ?

Reply:
Good point. I see that all the other already existing functions for other key types do it the same way. This does not necessarrily mean that the others do it right ;-)
These functions are called when creating a new object, which is either Import (MODE_CREATE) or Key generation (MODE_KEYGEN). In addition, they are called when copying an object (MODE_COPY).
For KEYGEN these attributes are possibly not required, because they are built and added by the key generation anyway. For COPY it could be assumed that the copied object already is valid, otherwise it would not exist. For IMPORT (CREATE), the attributes need to be checked, which is performed by the code. 
So the code 'might' be OK as is.....

Reply:
okay, so why not check the mode directly at the top and return in case != CREATE . I mean right now the code does everything to hide its doing essentially nothing in that case.